File: frontend/src/api/sheets.ts
import { apiGet, apiPost } from "./client";

export type DrawingSheet = {
  id: number;
  project_id: number;
  batch_id: number;
  file_id: number;
  page_no: number;
  sheet_type: string;
  drawing_no: string | null;
  drawing_name: string | null;
  discipline: string | null;
  version: string | null;
  issue_date: string | null;
  confidence_score: number | null;
  trust_level: string | null;
  status: string;
  review_status: string;
  thumbnail_path: string | null;
  preview_path: string | null;
  title_crop_path: string | null;
  title_crop_status: string | null;
  title_crop_error_code: string | null;
  title_crop_error_message: string | null;
  cad_preview_path: string | null;
  cad_preview_status: string | null;
  cad_preview_error_code: string | null;
  cad_preview_error_message: string | null;
  cad_preview_url: string | null;
  issue_count: number;
  error_count: number;
  warning_count: number;
  info_count: number;
  thumbnail_url: string | null;
  preview_url: string | null;
  original_file_name: string;
  source_format: string;
  created_at: string;
  updated_at: string;
};

export type PaginatedSheets = {
  items: DrawingSheet[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
};

export type SheetQuery = {
  keyword?: string;
  batch_id?: number;
  file_id?: number;
  discipline?: string;
  status?: string;
  review_status?: string;
  trust_level?: string;
  source_format?: string;
  issue_severity?: string;
  issue_code?: string;
  has_issue?: boolean;
  has_error?: boolean;
  has_warning?: boolean;
  low_confidence?: boolean;
  missing_field?: string;
  page?: number;
  page_size?: number;
  sort_by?: string;
  sort_order?: string;
};

export type BatchSplitResult = {
  batch_id: number;
  file_count: number;
  sheet_count: number;
  failed_count: number;
  files: Array<{
    file_id: number;
    page_count: number;
    sheet_count: number;
    failed_count: number;
  }>;
};

export function splitImportBatch(batchId: number): Promise<BatchSplitResult> {
  return apiPost<BatchSplitResult>(`/api/imports/${batchId}/split`, undefined, {
    timeoutMs: 300_000
  });
}

export function listProjectSheets(
  projectId: number,
  query: SheetQuery = {}
): Promise<PaginatedSheets> {
  const params = new URLSearchParams();
  Object.entries(query).forEach(([key, value]) => {
    if (value !== undefined && value !== "" && value !== null) {
      params.set(key, String(value));
    }
  });
  const queryString = params.toString();
  return apiGet<PaginatedSheets>(
    `/api/projects/${projectId}/sheets${queryString ? `?${queryString}` : ""}`
  );
}

export function getSheet(sheetId: number): Promise<DrawingSheet> {
  return apiGet<DrawingSheet>(`/api/sheets/${sheetId}`);
}

File: backend/api/sheets.py
from fastapi import APIRouter, Depends, Query
from fastapi.responses import FileResponse
from sqlalchemy.orm import Session

from backend.core.database import get_db
from backend.schemas.drawing_sheet import (
    BatchSplitResult,
    DrawingSheetListItem,
    DrawingSheetRead,
    FileSplitResult,
)
from backend.schemas.pagination import PaginatedResponse
from backend.schemas.sheet_query import SheetQueryParams
from backend.services import pdf_split_service, sheet_query_service

router = APIRouter(prefix="/api", tags=["sheets"])


@router.post("/files/{file_id}/split", response_model=FileSplitResult)
def split_file(file_id: int, db: Session = Depends(get_db)) -> FileSplitResult:
    return pdf_split_service.split_file(db, file_id)


@router.post("/imports/{batch_id}/split", response_model=BatchSplitResult)
def split_batch(batch_id: int, db: Session = Depends(get_db)) -> BatchSplitResult:
    return pdf_split_service.split_batch(db, batch_id)


@router.get("/projects/{project_id}/sheets", response_model=PaginatedResponse[DrawingSheetListItem])
def list_project_sheets(
    project_id: int,
    keyword: str | None = Query(default=None),
    file_id: int | None = Query(default=None),
    batch_id: int | None = Query(default=None),
    discipline: str | None = Query(default=None),
    status: str | None = Query(default=None),
    review_status: str | None = Query(default=None),
    trust_level: str | None = Query(default=None),
    source_format: str | None = Query(default=None),
    issue_severity: str | None = Query(default=None),
    issue_code: str | None = Query(default=None),
    has_issue: bool | None = Query(default=None),
    has_error: bool | None = Query(default=None),
    has_warning: bool | None = Query(default=None),
    low_confidence: bool | None = Query(default=None),
    missing_field: str | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    sort_by: str = Query(default="default"),
    sort_order: str = Query(default="asc"),
    db: Session = Depends(get_db),
) -> PaginatedResponse[DrawingSheetListItem]:
    return sheet_query_service.list_project_sheets(
        db,
        project_id,
        SheetQueryParams(
            keyword=keyword,
            file_id=file_id,
            batch_id=batch_id,
            discipline=discipline,
            status=status,
            review_status=review_status,
            trust_level=trust_level,
            source_format=source_format,
            issue_severity=issue_severity,
            issue_code=issue_code,
            has_issue=has_issue,
            has_error=has_error,
            has_warning=has_warning,
            low_confidence=low_confidence,
            missing_field=missing_field,
            page=page,
            page_size=page_size,
            sort_by=sort_by,
            sort_order=sort_order,
        ),
    )


@router.get("/sheets/{sheet_id}", response_model=DrawingSheetRead)
def get_sheet(sheet_id: int, db: Session = Depends(get_db)) -> DrawingSheetRead:
    return pdf_split_service.get_sheet_detail(db, sheet_id)


@router.get("/sheets/{sheet_id}/preview")
def get_sheet_preview(sheet_id: int, db: Session = Depends(get_db)) -> FileResponse:
    path = pdf_split_service.sheet_preview_path(db, sheet_id, "preview")
    return FileResponse(path, media_type="image/png")


@router.get("/sheets/{sheet_id}/thumbnail")
def get_sheet_thumbnail(sheet_id: int, db: Session = Depends(get_db)) -> FileResponse:
    path = pdf_split_service.sheet_preview_path(db, sheet_id, "thumbnail")
    return FileResponse(path, media_type="image/jpeg")
